// @crestron/ch5-shell-utilities-cli/build/cli/create-project/Ch5CreateProjectCli.js
"use strict";
// Copyright (C) 2022 to the present, Crestron Electronics, Inc.
// All rights reserved.
// No part of this software may be reproduced in any form, machine
// or natural, without the express written consent of Crestron Electronics.
// Use of this source code is subject to the terms of the Crestron Software License Agreement
// under which you licensed this source code.
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ch5CreateProjectCli = void 0;
var Ch5BaseClassForCliNew_1 = require("../Ch5BaseClassForCliNew");
var Ch5GeneratePageCli_1 = require("../generate-page/Ch5GeneratePageCli");
var Ch5GenerateWidgetCli_1 = require("../generate-widget/Ch5GenerateWidgetCli");
var path = require('path');
var fs = require("fs");
var fsExtra = require("fs-extra");
var process = require('process');
var editJsonFile = require("edit-json-file");
var Ch5CreateProjectCli = /** @class */ (function (_super) {
    __extends(Ch5CreateProjectCli, _super);
    /**
     * Constructor
     */
    function Ch5CreateProjectCli(showOutputMessages) {
        if (showOutputMessages === void 0) { showOutputMessages = true; }
        var _this = _super.call(this, "create-project") || this;
        _this.showOutputMessages = showOutputMessages;
        _this.SHELL_FOLDER = path.normalize(path.join(__dirname, "../../", "shell"));
        _this.PROJECT_CONFIG_JSON_PATH = path.normalize("/app/project-config.json");
        _this.VSCODE_SCHEMA_JSON_PATH = path.normalize(path.join(".vscode", "project-config-schema.json"));
        return _this;
    }
    /**
     * Initialize process
     */
    Ch5CreateProjectCli.prototype.initialize = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.logger.start("initialize");
                this.outputResponse.data.updateInputs = [];
                this.outputResponse.data.projectName = "";
                this.logger.end();
                return [2 /*return*/];
            });
        });
    };
    /**
     * Verify input parameters
     */
    Ch5CreateProjectCli.prototype.verifyInputParams = function () {
        return __awaiter(this, void 0, void 0, function () {
            var value, inputUpdate, tabDisplayText, warningMessage, i;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.logger.start("verifyInputParams");
                        if (!(this.inputArgs["config"].argsValue !== "")) return [3 /*break*/, 2];
                        // Step 1: Check file extension for json, valid input for 'config' argument, and config file existence
                        if (!(this.utils.isValidInput(this.inputArgs["config"].argsValue) && this.isConfigFileExist(this.inputArgs["config"].argsValue))) {
                            throw new Error(this.getText("VERIFY_INPUT_PARAMS.INVALID_CONFIG_INPUT"));
                        }
                        return [4 /*yield*/, this.isConfigFileValid(this.inputArgs["config"].argsValue, path.join(this.SHELL_FOLDER, this.VSCODE_SCHEMA_JSON_PATH), true)];
                    case 1:
                        // Step 2: Check if json is as per its schema (.vscode is hidden folder)
                        if (!(_a.sent())) {
                            throw new Error(this.getText("VERIFY_INPUT_PARAMS.INVALID_CONFIG_FILE"));
                        }
                        return [3 /*break*/, 3];
                    case 2:
                        this.logger.log("Blank project creation without json file");
                        Object.entries(this.inputArgs).forEach(function (_a) {
                            var key = _a[0], value = _a[1];
                            if (value.isSpecialArgument === false) {
                                var inputUpdate = __assign(__assign({}, value), { "warning": "" });
                                if (value.inputReceived === true) {
                                    if (_this.utils.isValidInput(value.argsValue)) {
                                        var validationResponse = _this.validateCLIInputArgument(value, value.key, value.argsValue, _this.getText("VERIFY_INPUT_PARAMS.INVALID_INPUT", value.key));
                                        _this.logger.log("validationResponse", validationResponse);
                                        if (validationResponse.warning === "") {
                                            inputUpdate.argsValue = validationResponse.value;
                                        }
                                        else {
                                            inputUpdate.argsValue = null;
                                            inputUpdate.warning = validationResponse.warning;
                                        }
                                    }
                                    _this.outputResponse.data.updateInputs.push(inputUpdate);
                                }
                            }
                        });
                        this.logger.log("this.outputResponse.data.updateInputs: ", this.outputResponse.data.updateInputs);
                        if (this.outputResponse.data.updateInputs.length === 0) {
                            value = this.inputArgs["projectName"];
                            inputUpdate = __assign(__assign({}, value), { "warning": "" });
                            inputUpdate.inputReceived = true;
                            this.outputResponse.data.updateInputs.push(inputUpdate);
                        }
                        this.logger.log("this.outputResponse.data.updateInputs: ", this.outputResponse.data.updateInputs);
                        tabDisplayText = this.getText("COMMON.HYPHEN_DELIMITER");
                        warningMessage = "";
                        for (i = 0; i < this.outputResponse.data.updateInputs.length; i++) {
                            if (this.utils.isValidInput(this.outputResponse.data.updateInputs[i].warning)) {
                                warningMessage += tabDisplayText + this.outputResponse.data.updateInputs[i].key + ": " + this.outputResponse.data.updateInputs[i].warning;
                            }
                        }
                        if (warningMessage !== "") {
                            if (this.showOutputMessages === true) {
                                this.logger.printWarning(this.getText("VERIFY_INPUT_PARAMS.INPUT_WARNING_TITLE", warningMessage));
                            }
                        }
                        _a.label = 3;
                    case 3:
                        this.logger.end();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Check if there are questions to be prompted to the integrator
     */
    Ch5CreateProjectCli.prototype.checkPromptQuestions = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _loop_1, this_1, i;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.logger.start("checkPromptQuestions");
                        if (!(this.inputArgs["config"].argsValue !== "")) return [3 /*break*/, 1];
                        return [3 /*break*/, 5];
                    case 1:
                        this.logger.log("this.outputResponse.data.updateInputs", this.outputResponse.data.updateInputs);
                        _loop_1 = function (i) {
                            var choicesList, componentsQuery, _b, questionsArray, response;
                            return __generator(this, function (_c) {
                                switch (_c.label) {
                                    case 0:
                                        if (!!this_1.utils.isValidInput(this_1.outputResponse.data.updateInputs[i].argsValue)) return [3 /*break*/, 4];
                                        if (!(this_1.outputResponse.data.updateInputs[i].type === "enum")) return [3 /*break*/, 2];
                                        choicesList = this_1.outputResponse.data.updateInputs[i].allowedValues;
                                        componentsQuery = new this_1.getSelect({
                                            name: 'value',
                                            message: this_1.getText(this_1.outputResponse.data.updateInputs[i].question),
                                            choices: choicesList
                                        });
                                        _b = this_1.outputResponse.data.updateInputs[i];
                                        return [4 /*yield*/, componentsQuery.run()
                                                .then(function (selectedMenu) { return selectedMenu; })
                                                .catch(function (error) { throw new Error(_this.getText("COMMON.SOMETHING_WENT_WRONG")); })];
                                    case 1:
                                        _b.argsValue = _c.sent();
                                        this_1.logger.log(this_1.outputResponse.data.updateInputs[i].key + ": ", this_1.outputResponse.data.updateInputs[i].argsValue);
                                        return [3 /*break*/, 4];
                                    case 2:
                                        if (!(this_1.outputResponse.data.updateInputs[i].type === "string")) return [3 /*break*/, 4];
                                        questionsArray = [
                                            {
                                                type: "text",
                                                name: this_1.outputResponse.data.updateInputs[i].key,
                                                initial: "shell-template",
                                                hint: "",
                                                message: this_1.getText(this_1.outputResponse.data.updateInputs[i].question),
                                                validate: function (inputValue) {
                                                    var valResponse = _this.validateCLIInputArgument(_this.outputResponse.data.updateInputs[i], _this.outputResponse.data.updateInputs[i]["key"], inputValue, _this.getText("VERIFY_INPUT_PARAMS.INVALID_INPUT", _this.outputResponse.data.updateInputs[i]["key"]));
                                                    if (valResponse.warning && valResponse.warning !== "") {
                                                        return valResponse.warning; // String output is considered as false for validate method.
                                                    }
                                                    else {
                                                        return true;
                                                    }
                                                }
                                            }
                                        ];
                                        return [4 /*yield*/, this_1.getEnquirer.prompt(questionsArray)];
                                    case 3:
                                        response = _c.sent();
                                        // Note that the prompt will keep asking inputs till it validates the correct value
                                        if (!this_1.utils.isValidInput(response.projectName)) {
                                            throw new Error(this_1.getText("COMMON.SOMETHING_WENT_WRONG"));
                                        }
                                        this_1.outputResponse.data.updateInputs[i].argsValue = response.projectName.toLowerCase();
                                        this_1.logger.log(this_1.outputResponse.data.updateInputs[i].key + ": ", this_1.outputResponse.data.updateInputs[i].argsValue);
                                        _c.label = 4;
                                    case 4: return [2 /*return*/];
                                }
                            });
                        };
                        this_1 = this;
                        i = 0;
                        _a.label = 2;
                    case 2:
                        if (!(i < this.outputResponse.data.updateInputs.length)) return [3 /*break*/, 5];
                        return [5 /*yield**/, _loop_1(i)];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4:
                        i++;
                        return [3 /*break*/, 2];
                    case 5:
                        this.logger.end();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Implement this component's main purpose
     */
    Ch5CreateProjectCli.prototype.processRequest = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projectFolderPath, newProjectConfigJSON, pathToCreateProject, isFolder, e_1, oldProjectConfigJSON, k, pagesToBeCreated, _loop_2, i, widgetsToBeCreated, _loop_3, i, packageJsonFile, i, genPage, newPageName, i, genWidget, newWidgetName, projectConfigJSON, pathToCreateProject, isFolder, e_2, packageJsonFile, genPage;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.logger.start("processRequest");
                        projectFolderPath = path.resolve(path.join("./"));
                        if (!(this.inputArgs["config"].argsValue !== "")) return [3 /*break*/, 14];
                        newProjectConfigJSON = JSON.parse(this.utils.readFileContentSync(this.inputArgs["config"].argsValue));
                        pathToCreateProject = path.resolve(path.join("./", newProjectConfigJSON.projectName));
                        if (!fs.existsSync(pathToCreateProject)) {
                            fs.mkdirSync(pathToCreateProject, { recursive: true });
                        }
                        this.logger.log("1. current working directory: " + process.cwd());
                        process.chdir(pathToCreateProject);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.utils.readdirAsync(pathToCreateProject)];
                    case 2:
                        isFolder = _a.sent();
                        if ((isFolder && isFolder.length > 0)) {
                            throw new Error(this.getText("PROCESS_REQUEST.FOLDER_CONTAINS_FILES", pathToCreateProject));
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _a.sent();
                        throw new Error(this.getText("PROCESS_REQUEST.FOLDER_CONTAINS_FILES", pathToCreateProject));
                    case 4:
                        fsExtra.copySync(this.SHELL_FOLDER, pathToCreateProject);
                        fs.renameSync(path.resolve(path.join(pathToCreateProject, "packagelock.json")), path.resolve(path.join(pathToCreateProject, "package-lock.json")));
                        this.logger.log("2. current working directory: " + process.cwd());
                        oldProjectConfigJSON = JSON.parse(this.utils.readFileContentSync(path.join(this.SHELL_FOLDER, this.PROJECT_CONFIG_JSON_PATH)));
                        // 1. Project Data
                        for (k in newProjectConfigJSON) {
                            if ((typeof newProjectConfigJSON[k] !== 'object' && newProjectConfigJSON[k] !== null)) {
                                if (oldProjectConfigJSON[k] !== newProjectConfigJSON[k]) {
                                    oldProjectConfigJSON[k] = newProjectConfigJSON[k];
                                    this.projectConfig.changeNodeValues(k, oldProjectConfigJSON[k]);
                                }
                            }
                        }
                        // 2. Themes
                        oldProjectConfigJSON["themes"] = newProjectConfigJSON["themes"];
                        this.projectConfig.changeNodeValues("themes", oldProjectConfigJSON["themes"]);
                        // 3. Config
                        oldProjectConfigJSON["config"] = newProjectConfigJSON["config"];
                        this.projectConfig.changeNodeValues("config", oldProjectConfigJSON["config"]);
                        // 4. Header
                        oldProjectConfigJSON["header"] = newProjectConfigJSON["header"];
                        this.projectConfig.changeNodeValues("header", oldProjectConfigJSON["header"]);
                        // 5. Footer
                        oldProjectConfigJSON["footer"] = newProjectConfigJSON["footer"];
                        this.projectConfig.changeNodeValues("footer", oldProjectConfigJSON["footer"]);
                        // 6. Content
                        oldProjectConfigJSON["content"]["triggerViewProperties"] = newProjectConfigJSON["content"]["triggerViewProperties"];
                        this.projectConfig.changeNodeValues("content.triggerViewProperties", oldProjectConfigJSON["content"]["triggerViewProperties"]);
                        oldProjectConfigJSON["content"]["$defaultView"] = newProjectConfigJSON["content"]["$defaultView"];
                        this.projectConfig.changeNodeValues("content.$defaultView", oldProjectConfigJSON["content"]["$defaultView"]);
                        pagesToBeCreated = [];
                        _loop_2 = function (i) {
                            var pageObj = newProjectConfigJSON["content"]["pages"][i];
                            var pageInOldSet = oldProjectConfigJSON["content"]["pages"].find(function (page) { return page.pageName.toString().toLowerCase() === pageObj.pageName.toString().toLowerCase(); });
                            if (!pageInOldSet) {
                                // Exists in New set but not in old - so create page
                                pagesToBeCreated.push(pageObj);
                            }
                        };
                        for (i = 0; i < newProjectConfigJSON["content"]["pages"].length; i++) {
                            _loop_2(i);
                        }
                        widgetsToBeCreated = [];
                        _loop_3 = function (i) {
                            var widgetObj = newProjectConfigJSON["content"]["widgets"][i];
                            var pageInOldSet = oldProjectConfigJSON["content"]["widgets"].find(function (widget) { return widget.widgetName.toString().toLowerCase() === widgetObj.widgetName.toString().toLowerCase(); });
                            if (!pageInOldSet) {
                                // Exists in New set but not in old - so create page
                                widgetsToBeCreated.push(widgetObj);
                            }
                        };
                        for (i = 0; i < newProjectConfigJSON["content"]["widgets"].length; i++) {
                            _loop_3(i);
                        }
                        packageJsonFile = editJsonFile("./package.json");
                        packageJsonFile.set("name", oldProjectConfigJSON.projectName);
                        packageJsonFile.save();
                        this.outputResponse.data.projectName = oldProjectConfigJSON.projectName;
                        i = 0;
                        _a.label = 5;
                    case 5:
                        if (!(i < pagesToBeCreated.length)) return [3 /*break*/, 8];
                        genPage = new Ch5GeneratePageCli_1.Ch5GeneratePageCli(false);
                        newPageName = pagesToBeCreated[i].fileName.toLowerCase().split(".")[0];
                        genPage.setInputArgsForTesting(["-n", newPageName, "-m", pagesToBeCreated[i].navigation ? "Y" : "N"]);
                        return [4 /*yield*/, genPage.run()];
                    case 6:
                        _a.sent();
                        this.projectConfig.replacePageNodeInJSON(pagesToBeCreated[i]);
                        _a.label = 7;
                    case 7:
                        i++;
                        return [3 /*break*/, 5];
                    case 8:
                        i = 0;
                        _a.label = 9;
                    case 9:
                        if (!(i < widgetsToBeCreated.length)) return [3 /*break*/, 12];
                        genWidget = new Ch5GenerateWidgetCli_1.Ch5GenerateWidgetCli(false);
                        newWidgetName = widgetsToBeCreated[i].fileName.toLowerCase().split(".")[0];
                        genWidget.setInputArgsForTesting(["-n", newWidgetName]);
                        return [4 /*yield*/, genWidget.run()];
                    case 10:
                        _a.sent();
                        this.projectConfig.removeWidgetFromJSON(widgetsToBeCreated[i]);
                        _a.label = 11;
                    case 11:
                        i++;
                        return [3 /*break*/, 9];
                    case 12: return [4 /*yield*/, this.isConfigFileValid(path.join(pathToCreateProject, this.PROJECT_CONFIG_JSON_PATH), path.join(pathToCreateProject, this.VSCODE_SCHEMA_JSON_PATH))];
                    case 13:
                        // Step 6: Run validate:project-config
                        if (!(_a.sent())) {
                            throw new Error(this.getText("PROCESS_REQUEST.PROJECT_CONFIG_VALIDATION_FAILED"));
                        }
                        // Step 8: Show proper messages  
                        this.outputResponse.result = true;
                        this.outputResponse.successMessage = this.getText("LOG_OUTPUT.SUCCESS_MESSAGE", this.outputResponse.data.projectName, projectFolderPath);
                        return [3 /*break*/, 21];
                    case 14:
                        projectConfigJSON = JSON.parse(this.utils.readFileContentSync(path.resolve(path.join(this.SHELL_FOLDER, this.PROJECT_CONFIG_JSON_PATH))));
                        // 1. Project Data
                        projectConfigJSON.projectName = this.outputResponse.data.updateInputs.find(function (objValue) { return objValue.key === "projectName"; }).argsValue;
                        pathToCreateProject = path.resolve(path.join("./", projectConfigJSON.projectName));
                        _a.label = 15;
                    case 15:
                        _a.trys.push([15, 17, , 18]);
                        return [4 /*yield*/, this.utils.readdirAsync(pathToCreateProject)];
                    case 16:
                        isFolder = _a.sent();
                        if ((isFolder && isFolder.length > 0)) {
                            throw new Error(this.getText("PROCESS_REQUEST.FOLDER_CONTAINS_FILES", pathToCreateProject));
                        }
                        return [3 /*break*/, 18];
                    case 17:
                        e_2 = _a.sent();
                        throw new Error(this.getText("PROCESS_REQUEST.FOLDER_CONTAINS_FILES", pathToCreateProject));
                    case 18:
                        if (!fs.existsSync(pathToCreateProject)) {
                            fs.mkdirSync(pathToCreateProject, { recursive: true });
                        }
                        fsExtra.copySync(this.SHELL_FOLDER, pathToCreateProject);
                        // Reason for below: https://docs.npmjs.com/cli/v8/configuring-npm/package-lock-json
                        fs.renameSync(path.resolve(path.join(pathToCreateProject, "packagelock.json")), path.resolve(path.join(pathToCreateProject, "package-lock.json")));
                        projectConfigJSON["content"]["$defaultView"] = "page1";
                        fs.writeFileSync(path.join(pathToCreateProject, this.PROJECT_CONFIG_JSON_PATH), JSON.stringify(projectConfigJSON));
                        process.chdir(pathToCreateProject);
                        packageJsonFile = editJsonFile("./package.json");
                        packageJsonFile.set("name", projectConfigJSON.projectName);
                        packageJsonFile.save();
                        this.outputResponse.data.projectName = projectConfigJSON.projectName;
                        genPage = new Ch5GeneratePageCli_1.Ch5GeneratePageCli(false);
                        genPage.setInputArgsForTesting(["-n", "page1", "-m", "Y"]);
                        return [4 /*yield*/, genPage.run()];
                    case 19:
                        _a.sent();
                        return [4 /*yield*/, this.isConfigFileValid(path.join(pathToCreateProject, this.PROJECT_CONFIG_JSON_PATH), path.join(pathToCreateProject, this.VSCODE_SCHEMA_JSON_PATH))];
                    case 20:
                        // Step 6: Run validate:project-config
                        if (!(_a.sent())) {
                            throw new Error(this.getText("PROCESS_REQUEST.PROJECT_CONFIG_VALIDATION_FAILED"));
                        }
                        this.outputResponse.result = true;
                        this.outputResponse.successMessage = this.getText("LOG_OUTPUT.SUCCESS_MESSAGE", this.outputResponse.data.projectName, projectFolderPath);
                        _a.label = 21;
                    case 21:
                        this.logger.end();
                        return [2 /*return*/];
                }
            });
        });
    };
    return Ch5CreateProjectCli;
}(Ch5BaseClassForCliNew_1.Ch5BaseClassForCliNew));
exports.Ch5CreateProjectCli = Ch5CreateProjectCli;
